cmake/optional.cmake

```diff
@@ -0,0 +1,106 @@
+# Copyright (c) 2023 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+# Optional features and packages.
+
+if(CCACHE)
+  find_program(CCACHE_EXECUTABLE ccache)
+  if(CCACHE_EXECUTABLE)
+    set(CCACHE ON)
+    if(MSVC)
+      # See https://github.com/ccache/ccache/wiki/MS-Visual-Studio
+      set(MSVC_CCACHE_WRAPPER_CONTENT "\"${CCACHE_EXECUTABLE}\" \"${CMAKE_CXX_COMPILER}\"")
+      set(MSVC_CCACHE_WRAPPER_FILENAME wrapped-cl.bat)
+      file(WRITE ${CMAKE_BINARY_DIR}/${MSVC_CCACHE_WRAPPER_FILENAME} "${MSVC_CCACHE_WRAPPER_CONTENT} %*")
+      set(CMAKE_VS_GLOBALS
+        "CLToolExe=${MSVC_CCACHE_WRAPPER_FILENAME}"
+        "CLToolPath=${CMAKE_BINARY_DIR}"
+        "TrackFileAccess=false"
+        "UseMultiToolTask=true"
+        "DebugInformationFormat=OldStyle"
+      )
```

Comment:
Apparently, this code does not work with cache 4.8 (

Going to investigate the issue, but I'm really hoping this won't be a blocker for this PR.

Reply:
ccache 4.8 is broken, ccache 4.7.4, 4.7.5 work fine.

Reply:
What/how is it broken in 4.8?

Reply:
Following https://github.com/ccache/ccache/wiki/MS-Visual-Studio literally for the simplest source code like "int main { return 0; }" works for 4.7, but 4.8 reports zero of "cacheable calls". 

For example, https://cirrus-ci.com/task/6670364843966464.

Reply:
Has this been reported upstream?

Reply:
> Has this been reported upstream?

https://github.com/ccache/ccache/issues/1262